File: Real-Time-Search-Workshop/dotnet/code/lab4.cs
/**
 * @fileoverview Advanced JSON operations with the node-redis client
 * @maker Joey Whelan
 */
using StackExchange.Redis;
using NRedisStack;
using NRedisStack.RedisStackCommands;
using System.Text.Json;

namespace SearchWorkshop 
{
    public class Lab4 
    {
        public void Run(IDatabase db) 
        {
            Console.WriteLine("\n*** Lab 4 - Data Loading ***");
            IJsonCommands json = db.JSON();
            json.Set("warehouse:1", "$", new {
                city = "Boston",
                location = "42.361145, -71.057083",
                inventory = new[] {
                    new {
                        id = 15970,
                        gender = "Men",
                        season = new[] {"Fall", "Winter"},
                        description = "Turtle Check Men Navy Blue Shirt",
                        price = 34.95
                    },
                    new {
                        id = 59263,
                        gender = "Women",
                        season = new[] {"Fall", "Winter", "Spring", "Summer"},
                        description = "Titan Women Silver Watch",
                        price = 129.99
                    },
                    new {
                        id = 46885,
                        gender = "Boys",
                        season = new[] {"Fall"},
                        description =  "Ben 10 Boys Navy Blue Slippers",
                        price = 45.99
                    }
                }
            });
    
            Console.WriteLine("\n*** Lab 4 - All Properties of Array ***");
            Console.WriteLine(json.Get(key: "warehouse:1",   
                path: "$.inventory[*]",
                indent: "\t", 
                newLine: "\n"
            ));

            Console.WriteLine("\n*** Lab 4 - All Properties of a Field ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[*].price", 
                indent: "\t", 
                newLine: "\n"
            ));

            Console.WriteLine("\n*** Lab 4 - Relational - Equality ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[?(@.description==\"Turtle Check Men Navy Blue Shirt\")]", 
                indent: "\t", 
                newLine: "\n"
            ));

            Console.WriteLine("\n*** Lab 4 - Relational - Less Than ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[?(@.price<100)]", 
                indent: "\t", 
                newLine: "\n"
            ));        

            Console.WriteLine("\n*** Lab 4 - Relational - Greater Than or Equal ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[?(@.id>=20000)]", 
                indent: "\t", 
                newLine: "\n"
            ));   

            Console.WriteLine("\n*** Lab 4 - Logical AND ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[?(@.gender==\"Men\"&&@.price>20)]", 
                indent: "\t", 
                newLine: "\n"
            ));  

            Console.WriteLine("\n*** Lab 4 - Logical OR ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[?(@.price<100||@.gender==\"Women\")].id", 
                indent: "\t", 
                newLine: "\n"
            ));  

            Console.WriteLine("\n*** Lab 4 - Regex - Contains Exact ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[?(@.description =~ \"Blue\")]", 
                indent: "\t", 
                newLine: "\n"
            )); 

            Console.WriteLine("\n*** Lab 4 - Regex - Contains, Case Insensitive ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[?(@.description =~ \"(?i)watch\")]", 
                indent: "\t", 
                newLine: "\n"
            )); 

            Console.WriteLine("\n*** Lab 4 - Regex - Begins With ***");
            Console.WriteLine(json.Get(key: "warehouse:1", 
                path: "$.inventory[?(@.description =~ \"^T\")]", 
                indent: "\t", 
                newLine: "\n"
            )); 
        }
    }
}

File: Real-Time-Query-Workshop/java/code/src/main/java/com/redis/queryworkshop/Lab4.java
/**
 * @fileoverview Advanced JSON operations with the redis client
 * @maker Joey Whelan
 */
package com.redis.queryworkshop;
import redis.clients.jedis.JedisPooled;
import redis.clients.jedis.json.Path;
import com.google.gson.Gson;

public class Lab4 {
    private class Product {
        private int id;
        private String gender;
        private String[] season;
        private String description;
        private double price;

        public Product(int id, String gender, String[] season, String description, double price) {
                this.id = id;
                this.gender = gender;
                this.season = season;
                this.description = description;
                this.price = price;
        }
    }

    private class Warehouse {
        private String city;
        private String location;
        private Product[] inventory;;

        public Warehouse(String city, String location, Product[] inventory) {
                this.city = city;
                this.location = location;
                this.inventory = inventory;
        }
    }

    public void run(JedisPooled client) {  
        System.out.println("\n*** Lab 4 - Data Loading ***");
        Product prod15970 = new Product(15970, "Men", new String[]{"Fall", "Winter"}, "Turtle Check Men Navy Blue Shirt",
            34.95);
        Product prod59263 = new Product(59263, "Women", new String[]{"Fall", "Winter", "Spring", "Summer"}, "Titan Women Silver Watch",
            129.99);
        Product prod46885 = new Product(46885, "Boys", new String[]{"Fall"}, "Ben 10 Boys Navy Blue Slippers",
            45.99);
        Warehouse wh1 = new Warehouse("Boston", "42.361145, -71.057083", 
            new Product[]{prod15970, prod59263, prod46885});
        Gson gson = new Gson();
        client.jsonSet("warehouse:1", gson.toJson(wh1));    
         
        System.out.println("\n*** Lab 4 - All Properties of Array ***");
        String res = client.jsonGetAsPlainString("warehouse:1", new Path("$.inventory[*]"));
        System.out.println(res);
   
        System.out.println("\n*** Lab 4 - All Properties of a Field ***");
        res = client.jsonGetAsPlainString("warehouse:1", new Path("$.inventory[*].price"));
        System.out.println(res);

        System.out.println("\n*** Lab 4 - Relational - Equality ***");
        res = client.jsonGetAsPlainString("warehouse:1", 
                new Path("$.inventory[?(@.description==\"Turtle Check Men Navy Blue Shirt\")]"));
        System.out.println(res);
 
        System.out.println("\n*** Lab 4 - Relational - Less Than ***");
        res = client.jsonGetAsPlainString("warehouse:1", new Path("$.inventory[?(@.price<100)]"));
        System.out.println(res);
       
        System.out.println("\n*** Lab 4 - Relational - Greater Than or Equal ***");
        res = client.jsonGetAsPlainString("warehouse:1", new Path("$.inventory[?(@.id>=20000)]"));
        System.out.println(res);      
 
        System.out.println("\n*** Lab 4 - Logical AND ***");
        res = client.jsonGetAsPlainString("warehouse:1", 
                new Path("$.inventory[?(@.gender==\"Men\"&&@.price>20)]"));
        System.out.println(res);

        System.out.println("\n*** Lab 4 - Logical OR ***");
        res = client.jsonGetAsPlainString("warehouse:1", 
                new Path("$.inventory[?(@.price<100||@.gender==\"Women\")].id"));
        System.out.println(res); 

        System.out.println("\n*** Lab 4 - Regex - Contains Exact ***");
        res = client.jsonGetAsPlainString("warehouse:1", 
                new Path("$.inventory[?(@.description =~ \"Blue\")]"));
        System.out.println(res);  

        System.out.println("\n*** Lab 4 - Regex - Contains, Case Insensitive ***");
        res = client.jsonGetAsPlainString("warehouse:1", 
                new Path("$.inventory[?(@.description =~ \"(?i)watch\")]"));
        System.out.println(res);  

        System.out.println("\n*** Lab 4 - Regex - Begins With ***");
        res = client.jsonGetAsPlainString("warehouse:1", 
            new Path("$.inventory[?(@.description =~ \"^T\")]"));
        System.out.println(res);  
    }
}
